src/palindrom-network-channel.js

```diff
@@ -0,0 +1,365 @@
+        }
+
+        onReceive && (this.onReceive = onReceive);
+        onSend && (this.onSend = onSend);
+        onConnectionError && (this.onConnectionError = onConnectionError);
+        onFatalError && (this.onFatalError = onFatalError);
+        onStateChange && (this.onStateChange = onStateChange);
+        onSocketOpened && (this.onSocketOpened = onSocketOpened);
+
+        Object.defineProperty(this, 'useWebSocket', {
+            get: function() {
+                return useWebSocket;
+            },
+            set: newValue => {
+                useWebSocket = newValue;
+
+                if (newValue == false) {
+                    if (this._ws) {
+                        this._ws.onclose = function() {
+                            //overwrites the previous onclose
+                            this._ws = null;
+                        };
+                        this._ws.close();
+                    }
+                    // define wsUrl if needed
+                } else if (!this.wsUrl) {
+                    this.wsUrl = toWebSocketURL(this.remoteUrl.href);
+                }
+                return useWebSocket;
+            }
+        });
+    }
+    get useWebSocket() {
+        return this._useWebSocket;
+    }
+    set useWebSocket(newValue) {
+        this._useWebSocket = newValue;
+
+        if (newValue == false) {
+            if (this._ws) {
+                this._ws.onclose = () => {
+                    //overwrites the previous onclose
+                    this._ws = null;
+                };
+                this._ws.close();
+            }
+            // define wsUrl if needed
+        } else if (!this.wsUrl) {
+            this.wsUrl = toWebSocketURL(this.remoteUrl.href);
+        }
+        return this.useWebSocket;
+    }
+
+    async _establish(reconnectionPendingData = null) {
+        const data = await this._xhr(
+            this.remoteUrl.href + (reconnectionPendingData ? '/reconnect' : ''),
+            'application/json',
+            reconnectionPendingData
+        );
+        if (this.useWebSocket) {
+            this.webSocketUpgrade(this.onSocketOpened);
+        }
+        return data;
+    }
+
+    /**
+     * Send any text message by currently established channel
+     * @TODO: handle readyState 2-CLOSING & 3-CLOSED (tomalec)
+     * @param  {String} msg message to be sent
+     * @return {PalindromNetworkChannel}     self
+     */
+    async send(msg) {
+        // send message only if there is a working ws connection
+        if (this.useWebSocket && this._ws && this._ws.readyState === 1) {
+            this._ws.send(msg);
+            this.onSend({ data: msg, url: this._ws.url, method: 'WS' });
+        } else {
+            const url = this.remoteUrl.href;
+            const data = await this._xhr(
+                url,
+                'application/json-patch+json',
+                msg
+            );
+            this.onReceive(data, url, 'GET');
```

Comment:
According to the line 318 in this file, the method can be GET  or PATCH.

Reply:
Actually, it's always `PATCH`.

Reply:
"always"? Howe are we guaranteed that there is always a `msg`? If `msg` is falsy, it would be `GET`

Reply:
I addressed it by removing method picking logic from `_xhr`, so it's more explicit and hopefully not to verbose, and ~D~RY